source/audio/Audio.cpp

```diff
@@ -355,179 +339,86 @@ void Audio::Step(bool isFastForward)
-	for(const Source &source : sources)
+	for(auto it = loopingPlayers.begin(); it != loopingPlayers.end();)
 	{
-		if(source.GetSound()->IsLooping())
+		const auto &[sound, player] = *it;
+		auto queueIt = soundQueue.find(sound);
+		if(queueIt != soundQueue.end())
 		{
-			auto it = soundQueue.find(source.GetSound());
-			if(it != soundQueue.end())
-			{
-				source.Move(it->second);
-				newSources.push_back(source);
-				soundQueue.erase(it);
-			}
-			else
-			{
-				alSourcei(source.ID(), AL_LOOPING, false);
-				endingSources.push_back(source.ID());
-			}
+			Move(*player, queueIt->second);
+			soundQueue.erase(queueIt);
+			++it;
 		}
 		else
 		{
-			// Non-looping sounds: check if they're done playing.
-			ALint state;
-			alGetSourcei(source.ID(), AL_SOURCE_STATE, &state);
-			if(state == AL_PLAYING || state == AL_PAUSED)
-				newSources.push_back(source);
-			else
-				recycledSources.push_back(source.ID());
+			reinterpret_cast<VolumeFadePlayer *>(player.get())->FadeOut();
+			it = loopingPlayers.erase(it);
 		}
 	}
-	// These sources were looping and are now wrapping up a loop.
-	auto it = endingSources.begin();
-	while(it != endingSources.end())
+
+	// Queue up the new buffers in every player, and remove the finished ones.
+	for(const auto &player : players)
 	{
-		ALint state;
-		alGetSourcei(*it, AL_SOURCE_STATE, &state);
-		if(state == AL_PLAYING)
-		{
-			// Fade out the sound. This avoids a clicking or rasping sound if a
-			// sound is cut off in the middle of its loop.
-			float gain = 1.f;
-			alGetSourcef(*it, AL_GAIN, &gain);
-			gain = max(0.f, gain - .05f);
-			alSourcef(*it, AL_GAIN, gain);
-			++it;
-		}
-		else if(state == AL_PAUSED)
-			++it;
-		else
-		{
-			recycledSources.push_back(*it);
-			it = endingSources.erase(it);
-		}
+		player->Supplier()->Set3x(isFastForward);
+		player->Update();
 	}
-	newSources.swap(sources);
+
+	erase_if(players, [](const auto &player){return player->IsFinished();});
 
 	// Now, what is left in the queue is sounds that want to play, and that do
 	// not correspond to an existing source.
-	for(const auto &it : soundQueue)
+	for(const auto &[sound, entry] : soundQueue)
 	{
-		// Use a recycled source if possible. Otherwise, create a new one.
-		unsigned source = 0;
-		if(recycledSources.empty())
+		unique_ptr<AudioSupplier> supplier = sound->CreateSupplier();
+		supplier->Set3x(isFastForward);
+		shared_ptr<AudioPlayer> player;
+		if(sound->IsLooping())
 		{
-			if(sources.size() >= maxSources)
-				break;
-
-			alGenSources(1, &source);
-			if(!source)
-			{
-				// If we just tried to generate a new source and OpenAL would
-				// not give us one, we've reached this system's limit for the
-				// number of concurrent sounds.
-				maxSources = sources.size();
-				break;
-			}
-		}
-		else
-		{
-			source = recycledSources.back();
-			recycledSources.pop_back();
+			player = shared_ptr<AudioPlayer>{new VolumeFadePlayer(entry.category, std::move(supplier))};
```

Comment:
I'm not asking for consistency across the whole PR, but in the same line it would be nice.
```suggestion
			player = shared_ptr<AudioPlayer>{new VolumeFadePlayer{entry.category, std::move(supplier)}};
```